agent/xds/listeners.go

```diff
@@ -71,16 +73,97 @@ func (s *Server) listenersFromSnapshotConnectProxy(cfgSnap *proxycfg.ConfigSnaps
 		}
 		resources[i+1] = upstreamListener
 	}
+
+	cfgSnap.Proxy.Expose.Finalize(s.Logger)
+	paths := cfgSnap.Proxy.Expose.Paths
+
+	// Add service health checks to the list of paths to create listeners for if needed
+	if cfgSnap.Proxy.Expose.Checks {
+		for _, check := range s.CheckFetcher.ServiceHTTPChecks(cfgSnap.Proxy.DestinationServiceID) {
+			p, err := parseCheckPath(check)
+			if err != nil {
+				s.Logger.Printf("[WARN] envoy: failed to create listener for check '%s': %v", check.CheckID, err)
+				continue
+			}
+			paths = append(paths, p)
+		}
+	}
+
+	// Configure additional listener for exposed check paths
+	for _, path := range paths {
+		clusterName := LocalAppClusterName
+		if path.LocalPathPort != cfgSnap.Proxy.LocalServicePort {
+			clusterName = makeExposeClusterName(path.LocalPathPort)
+		}
+
+		l, err := s.makeExposedCheckListener(cfgSnap, clusterName, path.Path, path.Protocol, path.ListenerPort)
```

Comment:
If there are multiple paths exposed on the same port, won't this create multiple listeners all bound to the same port which will fail? only the first one will be accepted by Envoy so the other paths won't be exposed? Or did I miss some part where we grouped paths by port already?

Reply:
This will fail if multiple listeners try to bind to the same port, so I should check for duplicate listener ports in the `NodeService.Validate()`.

Reply:
Added a check to `NodeService.Validate()` so we'll return an error if someone tries to provide duplicate listener ports.